### PR TITLE
Fix `Ipv6Addr::is_unicast_global` to check for unicast global scope

### PR DESCRIPTION
**Pull request by CDirkx**
See #85604 for previous discussion.

The current behaviour of `Ipv6Addr::is_unicast_global` is

```rust
/// Returns [`true`] if the address is a globally routable unicast address.
```

This is inconsistent with methods like `is_unicast_link_local` and `is_unicast_site_local`, which check if an address has unicast link-local scope or site-local scope respectively. Similarly one would expect `is_unicast_global` to check if an address has unicast global scope (which is not the same thing as being globally routable, for that we already have `Ipv6Addr::is_global` which is reworked in #86634).

This PR changes the behaviour to

```rust
/// Returns `true` if the address is a unicast address with global scope, as defined in [RFC 4291].
///
/// Any unicast address has global scope if it is not:
///  - the [unspecified address] (`::`)
///  - the [loopback address] (`::1`)
///  - a [link-local address] (`fe80::/10`)
```

Basing itself on the following table from [RFC 4291](https://datatracker.ietf.org/doc/html/rfc4291):

```
Address type         Binary prefix        IPv6 notation   Section
------------         -------------        -------------   -------
Unspecified          00...0  (128 bits)   ::/128          2.5.2
Loopback             00...1  (128 bits)   ::1/128         2.5.3
Multicast            11111111             FF00::/8        2.7
Link-Local unicast   1111111010           FE80::/10       2.5.6
Global Unicast       (everything else)
```

The documentation is also updated to reflect that future RFCs may change the behaviour of this method: "Future specifications may redefine one or more sub-ranges of the Global Unicast space for other purposes".

The behaviour of `is_global` is kept the same, it used to delegate to `is_unicast_global` but now performs the checks for global reachability itself.

To avoid further confusion between `is_unicast_global` and `is_global`, and the concepts of "global scope" and "globally reachable", the folowing methods have been renamed:
- `is_unicast_global` -> `has_unicast_global_scope`
- `is_unicast_link_local` -> `has_unicast_link_local_scope`
- <s>`is_unicast_link_local_strict` -> `has_unicast_link_local_scope_strict`</s> (removed in #85819)
- <s>`is_unicast_site_local` -> `has_unicast_site_local_scope`</s> (removed in #85820)

### CODE REVIEW

**Comment by the8472** [comment]
What does that mean for unique local addresses now? Those clearly are not global and RFC 4291 doesn't even acknowledge their  existence even though RFC 4193 predates it by a year.

**Comment by CDirkx** [comment]
Yes that is where the difference between "global scope" and "globally reachable" comes into play. RFC 4193 has the following:

> In practice, applications may treat these addresses like global scoped addresses.

> Application and other higher level protocols can treat [Unique Local] IPv6 addresses in the same manner as other types of global unicast addresses. No special handling is required. This type of address may not be reachable, but that is no different from other types of IPv6 global unicast address.

So unique local addresses are indeed not globally reachable (which is implemented in `is_global`) but do have global scope.

**Comment by CDirkx** [comment]
Perhaps it would be less confusing to rename `is_unicast_global` (and the link-local and site-local versions) to `has_unicast_global_scope` in order to distinguis it from `is_global`

**Comment by CDirkx** [comment]
I rebased and reordered the methods so now `has_unicast_global_scope` and `has_unicast_link_local_scope` are together.

File: library/std/src/net/ip.rs [hunk 1]
```diff
@@ -1239,7 +1239,10 @@ impl Ipv6Addr {
     pub const fn is_global(&self) -> bool {
         match self.multicast_scope() {
             Some(Ipv6MulticastScope::Global) => true,
-            None => self.is_unicast_global(),
+            None => {
+                self.has_unicast_global_scope()
+                    && !(self.is_unique_local() || self.is_documentation())
+            }
```
Comment (CDirkx) [inline library/std/src/net/ip.rs:1245]: This implementation of `is_global` is changed completely in #86634

**Comment by crlf0710** [comment]
Block on #86634

**Comment by chotchki** [comment]
@CDirkx I'm trying to figure out where the "ip" feature left off and I think its on this rebase need. Am I correct?

**Comment by Mark-Simulacrum** [comment]
Approved https://github.com/rust-lang/rust/pull/99947, so going to go ahead and close this PR. https://github.com/rust-lang/rust/pull/99947 contains the commits from here though.

**Pull request closed without merging**